File: transport/client/src/fec/LossEstimator.ts
export interface LossStats {
  lossRate: number;
  sampleSize: number;
}

const WINDOW_MS = 2000;

interface SentRecord {
  sequence: bigint;
  timestamp: number;
}

export class LossEstimator {
  private sentPackets: SentRecord[] = [];
  private nackedSet: Set<bigint> = new Set();
  private smoothedLossRate: number = 0;

  recordPacketSent(sequence: bigint): void {
    this.sentPackets.push({ sequence, timestamp: Date.now() });
    this.prune();
    this.updateSmoothedRate();
  }

  recordNackReceived(sequences: bigint[]): void {
    const windowSeqs = new Set(this.sentPackets.map(p => p.sequence));
    for (const seq of sequences) {
      if (windowSeqs.has(seq)) {
        this.nackedSet.add(seq);
      }
    }
    this.updateSmoothedRate();
  }

  getStats(): LossStats {
    return {
      lossRate: this.smoothedLossRate,
      sampleSize: this.sentPackets.length,
    };
  }

  private prune(): void {
    const cutoff = Date.now() - WINDOW_MS;
    const toRemove: bigint[] = [];

    this.sentPackets = this.sentPackets.filter(p => {
      if (p.timestamp < cutoff) {
        toRemove.push(p.sequence);
        return false;
      }
      return true;
    });

    for (const seq of toRemove) {
      this.nackedSet.delete(seq);
    }
  }

  private updateSmoothedRate(): void {
    if (this.sentPackets.length === 0) return;

    const raw = this.nackedSet.size / this.sentPackets.length;

    if (this.smoothedLossRate === 0) {
      this.smoothedLossRate = raw;
    } else if (raw > this.smoothedLossRate) {
      this.smoothedLossRate = 0.8 * this.smoothedLossRate + 0.2 * raw;
    } else {
      this.smoothedLossRate = 0.95 * this.smoothedLossRate + 0.05 * raw;
    }
  }

  reset(): void {
    this.sentPackets = [];
    this.nackedSet.clear();
    this.smoothedLossRate = 0;
  }
}

File: transport/server/src/fec/loss_estimator.rs
use std::collections::HashSet;
use std::time::{Duration, Instant};

#[derive(Debug, Clone)]
pub struct LossStats {
    pub loss_rate: f64,
    pub sample_size: usize,
}

const WINDOW_DURATION: Duration = Duration::from_millis(2000);

struct SentRecord {
    sequence: u64,
    timestamp: Instant,
}

pub struct LossEstimator {
    sent_packets: Vec<SentRecord>,
    nacked_set: HashSet<u64>,
    smoothed_loss_rate: f64,
}

impl LossEstimator {
    pub fn new() -> Self {
        Self {
            sent_packets: Vec::new(),
            nacked_set: HashSet::new(),
            smoothed_loss_rate: 0.0,
        }
    }

    pub fn record_packet_sent(&mut self, sequence: u64) {
        self.sent_packets.push(SentRecord {
            sequence,
            timestamp: Instant::now(),
        });
        self.prune();
        self.update_smoothed_rate();
    }

    pub fn record_nack_received(&mut self, sequences: &[u64]) {
        let window_seqs: HashSet<u64> = self.sent_packets.iter().map(|p| p.sequence).collect();
        for &seq in sequences {
            if window_seqs.contains(&seq) {
                self.nacked_set.insert(seq);
            }
        }
        self.update_smoothed_rate();
    }

    pub fn get_stats(&self) -> LossStats {
        LossStats {
            loss_rate: self.smoothed_loss_rate,
            sample_size: self.sent_packets.len(),
        }
    }

    fn prune(&mut self) {
        let cutoff = Instant::now() - WINDOW_DURATION;
        let mut to_remove = Vec::new();

        self.sent_packets.retain(|p| {
            if p.timestamp < cutoff {
                to_remove.push(p.sequence);
                false
            } else {
                true
            }
        });

        for seq in to_remove {
            self.nacked_set.remove(&seq);
        }
    }

    fn update_smoothed_rate(&mut self) {
        if self.sent_packets.is_empty() {
            return;
        }

        let raw = self.nacked_set.len() as f64 / self.sent_packets.len() as f64;

        if self.smoothed_loss_rate == 0.0 {
            self.smoothed_loss_rate = raw;
        } else if raw > self.smoothed_loss_rate {
            self.smoothed_loss_rate = 0.8 * self.smoothed_loss_rate + 0.2 * raw;
        } else {
            self.smoothed_loss_rate = 0.95 * self.smoothed_loss_rate + 0.05 * raw;
        }
    }

    pub fn reset(&mut self) {
        self.sent_packets.clear();
        self.nacked_set.clear();
        self.smoothed_loss_rate = 0.0;
    }
}

impl Default for LossEstimator {
    fn default() -> Self {
        Self::new()
    }
}
